TEP-0081: Add Chains sub-command to the CLI

rgreinho opened the pull request:
Adds a new TEP to propose to add a CLI command for Chains.

rgreinho commented:
/assign @pradeepitm12

dlorenc commented:
This is awesome to see! cc @priyawadhwa as well, we'll take a look from the chains perspective.

dlorenc commented on teps/0081-add-chains-subcommand-to-the-cli.md:
````diff
@@ -0,0 +1,326 @@
+Extract the Chains payload from a taskrun:
+
+```bash
+$ tkn chains payload cache-image-pipelinerun-r-qd6xw-build-trusted-jrj5v
````
I always worry a bit about printing signed payloads without verifying the signature from a UX perspective. Do you think signature verification is in scope here for the tkn cli?

If not, what do you think about something like `tkn chains unsafe-payload` just to make it abundantly clear that this hasn't been verified?

rgreinho replied:
Good point!

What about we leave the name of the sub-command to `payload` but we add a `--verify` flag which would verify the signature? The reason I am proposing to keep the sub-command name is because `tkn chains unsafe-payload --verify` sounds unclear (i.e. "asking to verify something which is unsafe" 🤔 ).

dlorenc replied:
Yeah that flag issue makes sense... I still don't love the idea of the default behavior being unsafe though. I usually prefer people to have to "opt-in" to the unsafe version.

What about maybe:

```shell
tkn chains payload --skip-verify
```

To skip it, but the default would be to verify?

rgreinho replied:
Yes, it makes complete sense 👍

rgreinho replied:
Another thought: doing so would make it a requirement to add `cosign` and `in-toto` as dependencies of the CLI, like what I described in the `Notes/Caveats` section. Is that an issue?

dlorenc replied:
We're working on a lightweight, stdlib only verification package for cosign that should work here.

pradeepitm12 replied:
>tkn chains payload --skip-verify

This makes a lot more sense to me.
Just one question, do we see any other sub-commands to be associated with the payload.
`tkn chains payload extract --skip-verify`

rgreinho replied:
As of now, no. I don't have any other use cases to do with the payload except extracting it and verifying it.

vdemeester replied:
Then it could be `payload-extract` subcommand, but I think I prefer `payload extract` 😅

dlorenc replied:
Note: there's a cosign library in progress now that should make the dependency tree much leaner. https://github.com/sigstore/cosign/issues/666

dlorenc commented on teps/0081-add-chains-subcommand-to-the-cli.md:
````diff
@@ -0,0 +1,326 @@
+Change the provenance format:
+
+```bash
+$ tkn chains format in-toto
````
I love this!

priyawadhwa replied:
+1 !

pradeepitm12 replied:
:+1:

priyawadhwa commented on teps/0081-add-chains-subcommand-to-the-cli.md:
```diff
@@ -0,0 +1,326 @@
+
+Available Commands:
+  payload     Print a Tekton chains' payload for a specific taskrun
+  signature   Print a Tekton chains' signature for a specific taskrun
```
We could add a `tkn chains config` command for modifying the configmap as you mentioned above! 

```
tkn chains config set <key> <value>
tkn chains config unset <key> <value>
```

could be options

rgreinho replied:
Good option! 

I am a bit hesitant (but not against idea!) since we are trying to keep the CLI simple and high level. Therefore this would potential fit in the "non-goals" category.

However, if we want to provide a simple way to modify any of the values described in the [documentation](https://github.com/tektoncd/chains/blob/main/docs/config.md), then this is definitely the way to go!

pradeepitm12 replied:
:+1:

bobcatfish commented:
fyi @tektoncd/cli-maintainers

pradeepitm12 commented on teps/0081-add-chains-subcommand-to-the-cli.md:
````diff
@@ -0,0 +1,326 @@
+Extract the signature from a taskrun:
+
+```bash
+$ tkn chains signature cache-image-pipelinerun-r-qd6xw-build-trusted-jrj5v
````
The above suggestions will work here also.
`tkn chains signature extract cache-image-pipelinerun-r-qd6xw-build-trusted-jrj5v`
This would help us to add more sub-commands like `tkn chains signature <action>`

vdemeester commented:
/assign

vdemeester commented on teps/0081-add-chains-subcommand-to-the-cli.md:
````diff
@@ -0,0 +1,326 @@
+Change the provenance format:
+
+```bash
+$ tkn chains format in-toto
````
Should we have a subcommand for *configuring* chains (in OpenShift we have `oc adm` for example). `tkn chains adm …` or even `tkn adm chains …` (which would mean we could/would also introduce similar features for other component such as pipeline, triggers, …)

rgreinho replied:
@priyawadhwa proposed `tkn chains config` which would do it.

We could add a "convention" stating that each command which would offer the ability to edit its related configuration must provide a `config` sub-command. That's kind of the reverse of what you are proposing, where all configuration sub-commands would be grouped under `tkn adm`. Both approaches make sense to me, although I like the config `tkn <command> config` idea a bit more.

vdemeester replied:
@rgreinho the only "trick" with `tkn <command> config` is that, it applies to `chains` well, but not for the rest, like `tkn task config` or `tkn pipeline config`, what does it do ? `pipeline`, `task`, `clustertask`, `pipelinerun` and `taskrun` are all subcommand that are "configured" together (`tektoncd/pipeline`). This is similar for the triggers subcommand for example.

rgreinho replied:
Ah I see! Then a `tkn adm chains` makes more sense after this explanation 👍

rgreinho commented:
I rebased the PR and added everybody's comment to the best I could.

To sum up:
* Extracting the payload should be secure by default, always verifying the signature, unless explicitly specified by the user
  * The Sigstore group is working on a lightweight `cosign` version that could be added into the CLI 
* The Chains configuration should be editable via the CLI. However it would be best to implement a dedicated administration command like `tkn adm` which will be part of another TEP.

Questions:
* When using the `in-toto` format, we will then need to also include in-toto as a dependency in order to be consistent. Can anyone confirm it is not an issue?
* Do we need to keep the `tkn chains format` subcommand? I think it would be a nice shortcut, but it would also duplicate an action that could eventually be done via `tkn adm`.
* Should I start a new TEP (now) for the `tkn adm` command?

vdemeester commented on teps/0081-add-chains-subcommand-to-the-cli.md:
```diff
@@ -0,0 +1,337 @@
+
+### Notes/Caveats (optional)
+
+Some commands would require adding extra dependencies. For instance we would
+need to either ensure `cosign` is installed on the system or add it as a
+dependency in order to implement the commands having to deal with the `cosign`
+keys and signatures.
+
+However, the Sigstore group is working on a lightweight version of `cosign` that
+could be more easily added as a dependency.
```
Note we could also just "bail" when cosign is not there, with a decent message for people to know where to look for installing it.

rgreinho replied:
That's a good point. However if we use `cosign` as an external binary instead of the library, I think we will have to store the data into files temporarily to be able to pass it to `cosign`, which is not ideal.

dlorenc replied:
I think the library approach is going to be the best one long term.

vdemeester commented on teps/0081-add-chains-subcommand-to-the-cli.md:
```diff
@@ -0,0 +1,337 @@
+Why should this TEP _not_ be implemented?
+-->
+
+## Alternatives
+
+The alternative is to use a combination of shell tools and to know which exact
+annotation or key to query/update. It was ruled out since it complicates
+the operations for no good reason. See the
+[Use Cases (optional)](use-cases-optional) section for some examples.
```
There is one other alternative that would be that the `chains` provide a `tkn-chains` binary, and with the "execution model" we have in `tkn`, it would appear as a subcommand. One downside of this, is that it wouldn't be available by default and would complicate a bit the "packaging part". I'd rather have the chains command in, and secure by default (if history teach us anything is that things not enable or shipped by default are less adopted 😓 )

dlorenc replied:
+1!

vdemeester reviewed (approved):
Small comments, but overall looks good 🎉

priyawadhwa commented:
> Do we need to keep the tkn chains format subcommand? I think it would be a nice shortcut, but it would also duplicate an action that could eventually be done via tkn adm.

I think it could be nice as well, since modifying the chains config is pretty messy command right now  :) It could be deprecated if `tkn adm` is added

rgreinho commented:
> Then it could be payload-extract subcommand, but I think I prefer payload extract 😅

@vdemeester if we use `payload extract`, that adds an extra level of command (i.e. a sub-sub-command). I think we should stick to 2 levels, like `tkn <command> <action> [parameter]...`. So `payload-extract` would fit this scheme.

skaegi commented:
/assign @pritidesai

rgreinho commented:
@pritidesai, do you have any remark of feedback that would prevent this TEP to be merged?

pritidesai commented:
Hey @rgreinho I promise to review the TEP this week 🥰

pritidesai commented:
Overall looks great, thanks @rgreinho for working on this proposal. This is very exciting ❤️ 

/approve

A couple of minor comments, but otherwise looks great 👍 

> Do we need to keep the tkn chains format subcommand? I think it would be a nice shortcut, but it would also duplicate an action that could eventually be done via tkn adm.
> Should I start a new TEP (now) for the tkn adm command?

Based on the discussion in the API WG last week, the `format` subcommand will be part of the overall `adm` command, right? If that's case, can we update the use case 3 listed in this TEP.

> The behaviour must be secure by default, therefore verifying the signature when extracting the payload. 

Is it possible that the signature verification can fail here? How are the signatures verified?

pritidesai commented:
Hey @pradeepitm12, we are very close to merge this. Based on the TEP approval process, we need approval from all the assignees. Please add your approval whenever you can. Thanks 🙏

rgreinho commented:
> Based on the discussion in the API WG last week, the `format` subcommand will be part of the overall `adm` command, right? If that's case, can we update the use case 3 listed in this TEP.

It will eventually, but @priyawadhwa mentioned that it could be nice to have this function here and then we could deprecate it if needed when the `adm` command is implemented.

> Is it possible that the signature verification can fail here? How are the signatures verified? 

I  think it would be possible for it to fail. This is part of a bigger discussion that I plan to have in the implementation phase. My idea is to add these new features in chains itself, then leverage them in the CLI. This design will be discussed as part of the chains WG meeting of Sept 30th 2021.

jerop commented:
/lgtm in API WG

dlorenc commented:
cc @pradeepitm12 can we get this merged?

pritidesai commented:
we discussed this in the API WG this morning, @vdemeester mentioned its ready to merge 🎉 

reapplying @jerop's lgtm

/lgtm

vdemeester commented:
🎉
